src/workload.rs

```diff
@@ -601,17 +741,25 @@ impl WorkloadStore {
         // include them in load balancing we do to Services.
         if status == HealthStatus::Healthy {
             for (vip, pl) in &w.virtual_ips {
-                let ip = vip.parse::<IpAddr>()?;
-                for port in &pl.ports {
-                    let service_sock_addr = SocketAddr::from((ip, port.service_port as u16));
-                    self.vips
-                        .entry(service_sock_addr)
-                        .or_default()
-                        .insert((wip, port.target_port as u16));
-                    self.workload_to_vip
-                        .entry(wip)
-                        .or_default()
-                        .insert((service_sock_addr, port.target_port as u16));
+                let vip = vip.parse::<IpAddr>()?;
+
+                let ep = Endpoint {
+                    address: wip,
+                    port: pl.into(),
+                };
+                if let Some(svc) = self.vips.get_mut(&vip) {
+                    svc.endpoints.insert(ep.address, ep);
+                    self.workload_to_vip.entry(wip).or_default().insert(vip);
+                } else {
+                    // Can happen due to ordering issues
+                    trace!("pod has VIP {vip}, but VIP not found");
+                    if let Some(ep_map) = self.staged_vips.get_mut(&vip) {
```

Comment:
nit: I think you can do something like `self.staged_vips.entry(vip).or_default().insert(ep)`

Reply:
implemented in https://github.com/istio/ztunnel/pull/486/commits/0189043ac3e0d065052dacfbfffd1bd9223b43d8